### PR TITLE
feat(Select): add "open", "toggleOpen" props

### LINKED ISSUE #552: Improve typings in Select
- Make separate type for renderControl and onOpenChange args
- Extend type for ref in renderControl args (I get error in class component with ref: `Type 'Ref<HTMLElement>' is not assignable to type 'LegacyRef<HTMLDivElement> | undefined'.`)

### LINKED ISSUE #553: Add force close method in Select
It's necessary to have some way of force closing an opened `SelectList`.

The case is that my component uses `Select` and I need to update `Select`'s `options` array on resize. It would be better to close the list before update.

Example of usage:
image
image

I use workaround with setting opened/closed status in onOpenChange callback and click outside of Select for now. However it's better to use some direct way.

### PR DESCRIPTION
**Pull request by imechoim**
1. Add controlled `open` prop.
2. `onOpenChange` will be called when open changes. (works for controlled and uncontrolled `open` prop)
3. introduce `toggleOpen`. Accept open value or inverts current.
4. `setOpen` marked as **deprecated**

Closes #553
Closes #552

### CODE REVIEW

File: src/components/Select/types.ts [hunk 1]
```diff
@@ -3,17 +3,20 @@ import {TextInputView, TextInputSize, TextInputPin} from '../TextInput';
 import {ControlGroupProps, ControlGroupOption, QAProps} from '../types';
 import {Option, OptionGroup} from './tech-components';
 
+export type RenderControl = (props: {
```
Comment (korvin89) [inline src/components/Select/types.ts:6]: This type is located in `index.ts`, I think it should have more unique name, maybe `SelectRenderControl`

File: src/components/utils/useSelect/useOpenState.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
+import React from 'react';
+import type {UseOpenProps} from './types';
+
+export const useOpenState = (props: UseOpenProps) => {
+    const [open, setOpenState] = React.useState(props.defaultOpen || false);
+    const isOpenControlled = React.useMemo(() => typeof props.open === 'boolean', [props.open]);
```
Comment (korvin89) [inline src/components/utils/useSelect/useOpenState.ts:6]: `useMemo` is redundant here. `typeof` operation pretty light I think

File: src/components/utils/useSelect/useOpenState.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    const {onOpenChange, onClose} = props;
+
+    React.useEffect(() => {
+        if (!isOpenControlled) return;
+
+        setOpenState(props.open as boolean);
```
Comment (korvin89) [inline src/components/utils/useSelect/useOpenState.ts:12]: I think this property should be required or should have default value

File: src/components/utils/useSelect/useOpenState.ts [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+    const toggleOpen = React.useCallback(() => {
+        if (open) {
+            setClose();
+        } else {
+            setOpen();
+        }
+    }, [setClose, open, setOpen]);
+
+    return {open, setOpen, setClose, toggleOpen};
```
Comment (korvin89) [inline src/components/utils/useSelect/useOpenState.ts:37]: `setOpen` changes looks a little bit major. Can we do it without changing current api? Maybe add new methods with different names and deprecate old setOpen

File: src/components/utils/useSelect/__tests__/useOpenState.test.tsx [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+            expect(result.current.toggleOpen).toEqual(expect.any(Function));
+        });
+        test('check defaultOpen prop', () => {
+            const {result} = renderUseOpenStateHook({defaultOpen: true});
+
+            expect(result.current.open).toBe(true);
+        });
+    });
+    describe('use funtions to change state', () => {
+        test('setOnep/setClose changes open value', () => {
```
Comment (korvin89) [inline src/components/utils/useSelect/__tests__/useOpenState.test.tsx:26]: funtions -> functions
setOnep -> setOpen

**Comment by korvin89** [comment]
Add this properties to readme file please

File: src/components/utils/useSelect/useOpenState.ts (hunk 4 above)
Comment (imechoim) [reply src/components/utils/useSelect/useOpenState.ts:37]: Right now there is only one usage place (inside useSelect hook).
I didn't change the items of useSelect return object just extend it.

old items: `open, setOpen`
new items: `setClose, toggleOpen`

File: src/components/Select/Select.tsx [hunk 6]
```diff
@@ -194,7 +195,8 @@ export const Select = React.forwardRef<HTMLButtonElement, SelectProps>(function
         inlineStyles.width = width;
     }
 
-    const {onFocus, onBlur} = useOnFocusOutside({enabled: open, onFocusOutside: handleClose});
+    const setClose = useCallback(() => toggleOpen(false), [toggleOpen]);
```
Comment (ValeraS) [inline src/components/Select/Select.tsx:198]: Please, use `React.useCallback` for consistency.

Comment (ValeraS) [inline src/components/Select/Select.tsx:198]: `handleClose`

File: src/components/utils/useSelect/useOpenState.ts [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+        },
+        [setOpenState, open, onOpenChange, props.open],
+    );
+
+    return {
+        open: openValue,
+        /**
+         * @deprecated use on onEnterKeyDown on Dialog component
+         */
+        setOpen: toggleOpen,
```
Comment (ValeraS) [inline src/components/utils/useSelect/useOpenState.ts:32]: I think, it should not be here but in useSelect.

File: src/components/utils/useSelect/useOpenState.ts [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+                if (newOpen !== props.open) {
+                    onOpenChange?.(newOpen);
+                }
+            } else if (newOpen !== open) {
+                setOpenState(newOpen);
+                onOpenChange?.(newOpen);
+            }
+        },
+        [setOpenState, open, onOpenChange, props.open],
+    );
```
Comment (ValeraS) [inline src/components/utils/useSelect/useOpenState.ts:25]:
```suggestion
    const isControlled = typeof props.open === 'boolean';
    const openValue = isControlled ? props.open : open;

    const toggleOpen = React.useCallback(
        (val?: boolean) => {
            const newOpen = typeof val === 'boolean' ? val : !openValue;
            if (newOpen !== openValue) {
                    onOpenChange?.(newOpen);
                    if (!isControlled) {
                        setOpenState(newOpen);
                    }
            }
        },
        [openValue, onOpenChange, isControlled],
    );
```

**Review by ValeraS — CHANGES_REQUESTED** [review]

**Review by ValeraS — APPROVED** [review]

**Review by korvin89 — APPROVED** [review]

**Pull request merged**